engine/access/rest/middlewares.go

```diff
@@ -0,0 +1,60 @@
+package rest
+
+import (
+	"net/http"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/rs/zerolog"
+)
+
+func NotImplementedHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Header().Set("Content-Type", "application/json; charset=UTF-8")
+		w.WriteHeader(http.StatusNotImplemented)
+	})
+}
+
+// LoggingMiddleware creates a middleware which adds a logger interceptor to each request to log the request method, uri,
+// duration and response code
+func LoggingMiddleware(logger zerolog.Logger) mux.MiddlewareFunc {
```

Comment:
Could we divide this logger to not wrap our request handler but instead have one middleware before the call (prepare all thins like start the timer, create new response writer and also create new request logger but also any additional boilerplate things repeated in each request) and one after that logs the time.

Reply:
I don't think wrapping the serverHTTP is avoidable. The middleware (unlike filters) build on each other and [their example ](https://github.com/gorilla/mux#middleware) also does the same.

Reply:
I think what @sideninja is telling is: 

Middleware logic in general is chain of middlewares:

- Do something, move to next handler

here it is more like: 

- Do Something, run next handler, Do something.

As long as Logging is the middleware there is not a big problem. But if there will be more middlewares, @sideninja's suggestion allows adding multiple logs (traces) to request, so you can see which middleware is lagging. 

If I understood correctly he suggests something like: 

`MWStartLogging, RequestHandler, MWLog`

Reply:
I was thinking we could/should create a logger instance in the first logging middleware that then gets passed down in the chain of middlewares and the final handler anyhow uses a common response method which can after writing a response also logs the finish time etc. This way we don't have to initialize loggers anywhere else and as @bluesign 
also mentioned other middlewares can potentially use the instance passed down in the chain with the context value.

Reply:
make sense..I have split the `LoggingMiddleware` to `RequestStart` and `RequestEnd` so that it avoids the Do Something, run next handler, Do something again pattern.